Expansion loop due to constraints (original report in #10382)

(this was originally reported by @lpw25 in #10382, but I do not think we have an issue tracking this yet. With keep-expansion (#11648), we may now have the means to fix this situation)

The following code, _Originally posted by @lpw25 in https://github.com/ocaml/ocaml/issues/10382#issuecomment-829443340_, causes an expansion loop in `Includecore.type_manifest` (which calls `Ctype.expand_head_no_link`):
```ocaml
 module type S = sig 
     type 'a t = 'a 
       constraint 'a = < m : r > 
     and r = (< m : r >) t 
   end 
  
 module type T = S with type 'a t = 'b constraint 'a = < m : 'b >
```

Changing the expansion to the linking version (`Ctype.expand_head`) is not sufficient: in this case, we repeatedly expand this type:
```
try_expand_head {id=2033;level=100000000;scope=0;marks=0;desc=
                 Texpand(
                  {id=2027;level=100000000;scope=0;marks=0;desc=
                   Tconstr(t/284,
                    [{id=2028;level=100000000;scope=0;marks=0;desc=
                      Tobject(
                       {id=2029;level=100000000;scope=0;marks=0;desc=
                        Tfield(m,Fpublic,
                         {id=2031;level=100000000;scope=0;marks=0;desc=
                          Tpoly(
                           {id=2032;level=100000000;scope=0;marks=0;desc=
                            Texpand({id=2027},r/282[1],[])},[])},
                        {id=2030;level=100000000;scope=0;marks=0;desc=Tnil})},
                       ref None)}],[t/284])},
                  r/282[1],[])}
```
which does not yield a different type. Note that because we are expanding `t`, and the `Texpand` is referencing `r`, we don't memorialize the fact that we have expanded `t`, due to path compression.

I am not sure if keeping multiple abbreviations in `Texpand` will prevent this loop.

However, I am wondering if `Ctype.expand_abbrev_gen` should check for forward progress -- for example by checking if `ty == ty'` before calling `link_expand` -- this prevents this particular loop (but not in principal mode!), but I'm not sure if it is correct.

Note: this code is already in the testsuite at https://github.com/ocaml/ocaml/blob/2e29548dfc89f3e182cb683db09da1ea7976e6d5/testsuite/tests/typing-modules/merge_constraint.ml#L255-L267, and causes a stack overflow
<!--
Welcome to OCaml's Issue tracker!

OCaml's developers use this tracker for bugs and feature requests only, rather
than user support.

If you have questions about *using* OCaml, please ask at
https://discuss.ocaml.org (more people read Discuss than this tracker, and
you'll get confirmation of whether you've really found a bug or need a new
feature).

If your error came from the OCaml package manager, opam, (messages beginning
`[ERROR] The compilation of ...`), please start at
https://github.com/ocaml/opam-repository/issues/new.

Some libraries and tools which used to be part of OCaml are now maintained
separately. Please post questions about Graphics, Num, camlp4, LablTk, CamlDBM
or OCamlTopWin on Discuss or on their respective issue trackers (see [README.adoc](https://github.com/ocaml/ocaml/blob/trunk/README.adoc#separately-maintained-components)
for a full list).
-->

--- a/Changes
+++ b/Changes
@@ -89,6 +89,11 @@ Working version
   inheritance in mutually recursive definition, immediate object inheritance.
   (Leo White, review by Florian Angeletti)
 
+- #14753, #14754: Prevent module constraints (`with type`) from introducing
+  cyclic types by re-checking well-foundedness when a module constraint is
+  used.
+  (Stefan Muenzel, report by Leo White, review by Jacques Garrigue)
+
 ### Code generation and optimizations:
 
 * #13919: optimize `lazy v` when `v` is a structured constant

--- a/testsuite/tests/typing-modules/merge_constraint.ml
+++ b/testsuite/tests/typing-modules/merge_constraint.ml
@@ -262,8 +262,12 @@ module type T = S with type 'a t = 'b constraint 'a = < m : 'b >;;
 [%%expect{|
 module type S =
   sig type 'a t = 'a constraint 'a = < m : r > and r = < m : r > t end
-Uncaught exception: Stack overflow
-
+Line 7, characters 23-64:
+7 | module type T = S with type 'a t = 'b constraint 'a = < m : 'b >;;
+                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The definition of "t" contains a cycle:
+         "< m : r > t/2" = "r",
+         "r" = "< m : r > t/2"
 |}]
 
 (* Correct *)

--- a/typing/typedecl.mli
+++ b/typing/typedecl.mli
@@ -60,6 +60,10 @@ val approx_type_decl:
     explanation:Types.type_origin -> Parsetree.type_declaration list ->
                                   (Ident.t * type_declaration) list
 
+val check_well_founded_decl:
+  abs_env:Env.t ->final_env:Env.t -> is_decl_path:(Path.t -> bool) ->
+  Location.t -> Path.t -> Types.type_declaration -> unit
+
 (** [check_recmod_typedecl ~abs_env env loc recmod_ids path decl]
    - [recmod_ids] is the list of recursively-defined module idents.
    - [path, decl] is the type declaration to be checked.

--- a/typing/typemod.ml
+++ b/typing/typemod.ml
@@ -275,6 +275,11 @@ let check_type_decl env sg loc id row_id newdecl decl =
     let abs_ty = abstractify_type newdecl in
     Env.add_local_constraint path abs_ty env
   in
+  (* While any "with" constraints that could be used to create ill-founded
+     types in downstream definitions that depend on this one should be rejected
+     by the inclusion check below, we still need to ensure that the type
+     declarations input to the inclusion check are well-founded; otherwise
+     the inclusion check may loop on type declarations that it should reject. *)
   Typedecl.check_well_founded_decl
     ~abs_env ~final_env:env ~is_decl_path:(Path.same path) loc path
     newdecl;

--- a/typing/typemod.ml
+++ b/typing/typemod.ml
@@ -199,23 +199,23 @@ let type_module_type_of_fwd :
 (* Additional validity checks on type definitions arising from
    recursive modules *)
 
+let abstractify_type ty =
+  let arity = ty.type_arity in
+  { ty with
+    type_params =
+      List.map (fun _ -> Btype.newgenvar()) ty.type_params;
+    type_kind = Type_abstract Rec_check_regularity;
+    type_manifest = None;
+    type_variance = Variance.unknown_signature ~injective:false ~arity;
+    type_separability = Types.Separability.default_signature ~arity;
+    type_is_newtype = false;
+    type_expansion_scope = Btype.lowest_level;
+    type_immediate = Unknown;
+    type_unboxed_default = false;
+  }
+
 let check_recmod_typedecls env decls =
   let recmod_ids = List.map fst decls in
-  let abstractify_type ty =
-    let arity = ty.type_arity in
-    { ty with
-      type_params =
-        List.map (fun _ -> Btype.newgenvar()) ty.type_params;
-      type_kind = Type_abstract Rec_check_regularity;
-      type_manifest = None;
-      type_variance = Variance.unknown_signature ~injective:false ~arity;
-      type_separability = Types.Separability.default_signature ~arity;
-      type_is_newtype = false;
-      type_expansion_scope = Btype.lowest_level;
-      type_immediate = Unknown;
-      type_unboxed_default = false;
-    }
-  in
   let type_to_abstract =
     List.fold_left
       (fun acc (id, md) ->
@@ -271,6 +271,13 @@ let check_type_decl env sg loc id row_id newdecl decl =
     | Some fresh_row_id -> Env.add_type ~check:false fresh_row_id newdecl env
   in
   let env = Env.add_signature sg env in
+  let abs_env =
+    let abs_ty = abstractify_type newdecl in
+    Env.add_local_constraint path abs_ty env
+  in
+  Typedecl.check_well_founded_decl
+    ~abs_env ~final_env:env ~is_decl_path:(Path.same path) loc path
+    newdecl;
   Includemod.type_declarations ~mark:true ~loc env fresh_id newdecl decl;
   Typedecl.check_coherence env loc path newdecl
 

